jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Callback.java

```diff
@@ -389,6 +391,158 @@ static Callback from(Callback callback1, Callback callback2)
+            {
+                Throwable failure = state.get(h);
+                int s = h[0];
+
+                if (s >= 0)
+                    throw new IllegalStateException("closed");
+                // we can only create new callbacks when not closed, so we decrement to make a more negative stamp.
+                if (!state.compareAndSet(failure, failure, s--, s))
+                    continue;
+
+                return new Callback()
+                {
+                    private final AtomicBoolean completed = new AtomicBoolean(false);
+
+                    @Override
+                    public void failed(Throwable x)
+                    {
+                        if (completed.compareAndSet(false, true))
+                            complete(x);
+                    }
+
+                    @Override
+                    public void succeeded()
+                    {
+                        if (completed.compareAndSet(false, true))
+                            complete(null);
+                    }
+
+                    @Override
+                    public InvocationType getInvocationType()
+                    {
+                        return Invocable.getInvocationType(andThen);
+                    }
+                };
+            }
+        }
+
+        /**
+         * Called to indicate that no more calls to {@link #newCallback()} will happen.
+         * If the combination is already complete, then it will be completed in the scope of this call.
+         *
+         * @throws IllegalStateException if this method has already been called.
+         */
+        @Override
+        public void close() throws IllegalStateException
+        {
+            int[] h = new int[1];
+            while (true)
+            {
+                Throwable failure = state.get(h);
+                int s = h[0];
+
+                if (s >= 0)
+                    throw new IllegalStateException("closed");
+
+                // make the stamp positive to indicate that the combination is closed
+                if (!state.compareAndSet(failure, failure, s, -s))
+                    continue;
+                complete(null);
+                return;
+            }
+        }
+
+        private void complete(Throwable failed)
+        {
+            Throwable failure;
+            int[] h = new int[1];
+            while (true)
+            {
+                failure = state.get(h);
+                int s = h[0];
+
+                // combine failures
+                Throwable combined = failure;
+                if (combined == null)
+                    combined = failed;
+                else if (failed != null)
+                    ExceptionUtil.addSuppressedIfNotAssociated(failure, failed);
+
+                // If the stamp is < 0, the combination has not been closed, so we increment, else we decrement.
+                int n = s < 0 ? s + 1 : s - 1;
+                if (!state.compareAndSet(failure, combined, s, n))
```

Comment:
If this CAS fails, is exception combination above safe?

Reply:
It is.  Failure will only go from null to non-null once and then stay at that value.   So we are not trying to hit a moving target. If the cas fails, then the addSuppressed will not add the same exception again, because they will already be associated. But it will be the same target failure, so the previous addSuppressed was the correct thing to do.

Note, it is because of this slight complex race that I had the failure logic after the CaS, but before the the `andThen` callback. I think it is safe there as well with less thought needed, but you asked me to move it.  I think it is probably better where it is now.